Fail lookups on invalid resource names

rob-clarke opened the pull request:
Depends on https://github.com/ros2/ros2cli/pull/643

---

This was found through ros2launch, where attempting to provide an absolute path to a launch file, along with launch arguments, resulted in failure. This was because ros2launch was passing the absolute path to `get_package_prefix()` in attempting to see if it was a package.

If a valid absolute path was passed to `get_package_prefix()` as the package name, a `PackageNotFoundError` was **not** raised and instead the first path from `get_search_paths()` was returned.

This resulted from the behaviour of `get_resource()`: if a valid absolute path was passed to `get_resource()` as the resource name, the contents of that absolute path was returned along with the first path from `get_search_paths()` as the path.

This PR adds an `InvalidResourceNameError` which is raised when the resource name argument contains a `/`. This catches the absolute path case as well as cases that may result from relative paths. I couldn't find any hard requirements on resource names other than a suggestion [here](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md#marker-files) that they should conform to [REP 127](https://www.ros.org/reps/rep-0127.html#name) suggestions on naming, which are now captured in [REP 144](https://www.ros.org/reps/rep-0144.html). If stricter restrictions on resource names are defined, these can be added to the `name_is_invalid()` function added here.

I'm unsure if this is a good fit for here or the problem is better off being addressed in ros2launch, where input to `get_package_prefix()` could be sanitised. Even if changed here, a fix in ros2launch will be needed but the error message should now be a little clearer.

rob-clarke committed:
Fail lookups on absolute paths

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Update for linter errors

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Hopefully final linter fix

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

audrow reviewed (commented):
@rob-clarke, thanks for the PR the code and tests look good to me. 

As you pointed out, I'm not sure if this should be fixed in ros2launch. @hidmic, what do you think?

hidmic reviewed (commented):
> I couldn't find any hard requirements on resource names other than a suggestion here that they should conform to REP 127 suggestions on naming, which are now captured in REP 144.

@rob-clarke to the best of my knowledge, there aren't any strict requirements. I'd rather enforce valid package names instead of coming up with extra checks for resource names.

rob-clarke commented:
The current resource index [documentation](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md#marker-files) says:
> Implementations should consider that spaces are allowed in marker file names...

So I'm hesitant to enforce package names as it doesn't fit with the warnings there.

Currently the PR only checks for forward slashes in the resource name which fixes the issue that led to this. It could be made stricter to only accept valid package names and additionally spaces to follow the resource index doc.

hidmic commented:
@rob-clarke I meant enforcing valid package names where a package name is expected i.e. in `get_package_prefix()`, not in `get_resource`. Sorry if I wasn't clear.

rob-clarke commented:
Yes, a check for a valid package name can be added to `get_package_prefix()`. Would this then raise `PackageNotFound` or be more descriptive? _e.g._ raise `ValueError`

I think some form of resource name check is still necessary. Ultimately the cause of this was that if a resource name starting with `/` (or presumably `\` on Windows) is passed to `get_resource()`, the use of `os.path.join()` results in that name replacing the entire path. If that name is also a valid file path, the subsequent `os.path.isfile()` check passes and no exception is raised. As far as I can tell, resources are only ever expected to be marker _files_, so raising an exception for any resource name containing either `/` or `\` should be safe.

hidmic commented:
> Would this then raise PackageNotFound or be more descriptive? e.g. raise ValueError

`ValueError` seems appropriate.

> I think some form of resource name check is still necessary. Ultimately the cause of this was that if a resource name starting with / (or presumably \ on Windows) is passed to get_resource(), the use of os.path.join() results in that name replacing the entire path. If that name is also a valid file path, the subsequent os.path.isfile() check passes and no exception is raised. As far as I can tell, resources are only ever expected to be marker files, so raising an exception for any resource name containing either / or \ should be safe.

That's fair. I think it's reasonable to assume absolute paths are no-go for resource names and types, as these break search paths. But a relative path for a resource type or name wouldn't -- and it may be even be handy for someone, though I don't know of an existing use case.

Checking for invalid package names and absolute paths in both resource types and names should address the issue in `ros2 launch` when a path (absolute or relative) to a launch file is provided. WDYT?

rob-clarke committed:
Add valid package name check

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Linting

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke commented:
Quoting from the [docs](https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md#resource-types) again:
> Resource types are represented as folders in the resource index and should be shallow, i.e. there should only be marker files within the resource type folders. This means that there is no nesting of resource types, which keeps the file system flat, and makes answering "What resource types are in this <prefix>?" as easy as simply listing the directories in the resource index folder. Any folders within the resource type folders, and any folders/files starting with a dot (.), should be ignored while listing the directories in the resource index folder. Instead of nesting resource type folders, the convention of using prefixes and suffixes separated by periods should be used.

That suggests that relative paths should not be expected for resource type names or resource names.

Looking at the bigger picture: is looking up a resource of a type with a relative path name not just equivalent to looking up a different resource type? At which point would the user not be better off just calling `get_resource('type_i_actually_want','name')`  rather than gaming the API?

rob-clarke committed:
Add backslash to invalid chars for resource name

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

hidmic commented:
> That suggests that relative paths should not be expected for resource type names or resource names.

Indeed. I missed that piece of documentation. I'm fine with blocking slashes for resource types and names then.

> is looking up a resource of a type with a relative path name not just equivalent to looking up a different resource type?

Yes, it'd equivalent from a user POV. The ambiguity is not necessarily bad, but it appears it wasn't intended.

hidmic commented on ament_index_python/ament_index_python/resources.py:
```diff
@@ -34,6 +44,9 @@ def get_resource(resource_type, resource_name):
     assert resource_name, 'The resource name must not be empty'
+    if name_is_invalid(resource_name):
+        raise InvalidResourceNameError(
+            "Resource name '%s' is invalid" % resource_name)
```
@rob-clarke based on your findings, considering validating resource types as well.

wjwwood replied:
I think this was resolved. (marking it resolved)

hidmic commented on ament_index_python/ament_index_python/resources.py:
```diff
@@ -19,6 +19,16 @@
+
+
+def name_is_invalid(resource_name):
+    if ('/' in resource_name) or ('\\' in resource_name):
+        return True
+    return False
```
@rob-clarke nit: 
```suggestion
    return ('/' in resource_name) or ('\\' in resource_name)
```

hidmic reviewed (approved):
Overall LGTM

rob-clarke committed:
Add validity check for type

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

audrow reviewed (approved):
Looks good to me with green CI.

audrow commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14393)](http://ci.ros2.org/job/ci_linux/14393/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9167)](http://ci.ros2.org/job/ci_linux-aarch64/9167/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12068)](http://ci.ros2.org/job/ci_osx/12068/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14492)](http://ci.ros2.org/job/ci_windows/14492/)

clalancette commented:
While this is definitely a bugfix, I'm nervous about putting it into Galactic at this juncture.  It's in such a low-level package that any problems here have the potential to cause instabilities all the way up the stack.  I'm going to suggest that we hold off on this one for post-Galactic, and then consider backporting it into Galactic in a patch release.  If we go that route, I think this also deserves an entry in the Galactic release notes as a Known Issue.

audrow commented:
> I'm going to suggest that we hold off on this one for post-Galactic, and then consider backporting it into Galactic in a patch release.

Sounds good to me.

> If we go that route, I think this also deserves an entry in the Galactic release notes as a Known Issue.

If that is what we decide to do, I can add this to the release notes.

clalancette commented:
This one sounds like a good one to add to the (not-yet-created) Galactic Patch Release 1 board.  @cottsay please add this one as well when you get a chance.

rob-clarke commented:
The main thing to watch out for in the full stack testing is use of `get_package_prefix` with autogenerated package names that may not conform to [REP-144](https://www.ros.org/reps/rep-0144.html). The check implemented here checks for REP-144 but additionally allows consecutive underscores that may appear in autogenerated names:

> - not use multiple _ separators consecutively. This allows generated symbols to use the __ separator to guarenteed the avoidance of collisions with symbols from other packages, for example in the message generators.

audrow commented:
@rob-clarke, from you fix the CI errors? It looks like (1) there is a linter warning and (2) that another PR needs to be added to update a test in [`ros2pkg`](https://github.com/ros2/ros2cli/blob/2f0ed235ac4b34c20ec1ec93aa6d828e887acec2/ros2pkg/test/test_api.py#L32) to expect a `ValueError`. Let me know if you have any questions.

cottsay commented:
> This one sounds like a good one to add to the (not-yet-created) Galactic Patch Release 1 board. @cottsay please add this one as well when you get a chance.

Done: https://github.com/orgs/ros2/projects/32

rob-clarke committed:
Fix quote style

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Fix import ordering

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke commented:
I think that fixes the linter warnings.

I'll need a bit more input on the `ros2pkg` error. It comes down to deciding if a leading underscore is grounds for raising a `ValueError` for an invalid package name. By REP-144 it would be an invalid name, but if this is likely to be used with autogenerated package names that may not conform, it could be made less strict here.

clalancette commented:
@rob-clarke You've mentioned auto-generated package names that don't conform to REP-144 a couple of times here.  Could you explain more on where you expect these to come from?

rob-clarke commented:
It's not really that I expect them to exist, just that the standard leaves that option open so I don't want to break things elsewhere in the stack.

tfoote commented:
We should not relax the naming rules in REP 144.

First, we should not relax the REP for "generated" packages. To a user a generated package is the same as any other package on their system. You'll find that several packages already have generated content pragmatically. To keep "generated" packages from colliding I recommend using a unique prefix with the name relating to the project or organization. Packages operate in a flat namespace and it's the developer's responsibility not to clobber others in that namespace. Adding the prefix underscore will keep you from colliding with others not using the double underscore. But if you're generating packages without controls or limits it would seem to be easy for  `_foo` to be generated by two different generators which will collide because you'll never know when someone else is also "generating" packages.

And second, the name rules are leveraged in our different language code generators so that generated code cannot collide with user written code that follows our naming conventions. Relaxing the requirement that the underscores are only internal separators would prevent the generated code from using a double underscore to provide a higher level separator. This is particularly important as we need to have this same guarantee even if a new language support is added that the developer didn't know about when writing their code etc.

wjwwood commented on ament_index_python/ament_index_python/packages.py:
```diff
@@ -44,7 +45,11 @@ def get_package_prefix(package_name):
     :raises: :exc:`PackageNotFoundError` if the package is not found
+    :raises: :exc:`ValueError` if the package name is invalid
     """
+    if re.fullmatch('[a-z][a-z0-9_]+', package_name, re.ASCII) is None:
```
I guess it's not convenient to use an already existing check for this, e.g. from `ament_package` or something? (could very well be the case that it's not worth the effort)

If not, then I think we should have a back reference to REP 144 (or 127, which ever is appropriate) here as a comment. Otherwise this regex seems to have been plucked out of no where, no matter how obvious is it seems to us, it might be confusing in the future.

rob-clarke replied:
Closest thing I could find for an existing check is [in `catkin_pkg`](https://github.com/ros-infrastructure/catkin_pkg/blob/ad07ab2570006a72994657ab3f3267a9e84469aa/src/catkin_pkg/package.py#L231-L238). The match differs a little there as it includes the dashes and uppercase letters from REP-127's ["recommended exemptions"](https://ros.org/reps/rep-0127.html#name). It's not really in a reusable form there.

Dashes should be added to this match too to allow for the full range of package names that might show up from a manifest.

rob-clarke replied:
Changed to match REP-127 and `catkin_pkg`. Added a comment explaining source

wjwwood commented on ament_index_python/ament_index_python/resources.py:
```diff
@@ -19,6 +19,18 @@
+
+
+def name_is_invalid(resource_name):
+    return ('/' in resource_name) or ('\\' in resource_name)
```
It would be nice to have basic doc strings for these items (even if they are trivial).

rob-clarke replied:
Added

wjwwood commented on ament_index_python/ament_index_python/resources.py:
```diff
@@ -19,6 +19,18 @@
+    pass
+
+
+def name_is_invalid(resource_name):
```
Should we make any effort to make this private (not part of the public API)? If we imagine people might use it as part of the public API, then I'd recommend a more descriptive name, and more thorough documentation that rationalizes the checks, because without the context of this issue/pr discussion it might be hard figure out why this check makes these assertions specifically. A comment linking to this discussion or a summary of this discussion may be sufficient.

rob-clarke replied:
I don't think it's likely to be used publicly. I've added some more documentation anyway.

As it's not in the `__init__.py` it's not particularly public at the moment. Is adding a leading underscore to the function name sufficient to mark it as definitely private?

wjwwood replied:
I think that's what we've done in the past, and I think that's fine. We can also just say in the doc block that we don't consider it public API.

rob-clarke committed:
Update package name regex to match REP-127

Now allows uppercase letters, dashes and leading numerals.
Test modified to reflect this

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Add docstrings

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Add private API marker

Signed-off-by: Rob Clarke <robert.clarke@bristol.ac.uk>

rob-clarke committed:
Merge branch 'master' into fail-abspaths

audrow commented:
@rob-clarke, is this ready to be reviewed again by @wjwwood ?

rob-clarke commented:
Yes, I think I've addressed all the comments.

wjwwood reviewed (approved):
Lgtm

audrow commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14714)](http://ci.ros2.org/job/ci_linux/14714/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9466)](http://ci.ros2.org/job/ci_linux-aarch64/9466/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12381)](http://ci.ros2.org/job/ci_osx/12381/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14859)](http://ci.ros2.org/job/ci_windows/14859/)

audrow commented:
It seems that this test is now throwing a `ValueError`: https://github.com/ros2/ros2cli/blob/fe074a30979c18f148b119c008a4a49c11c346cf/ros2pkg/test/test_api.py#L32

@rob-clarke, if you make a small PR in `ros2cli`, you can ping me for review.

audrow commented:
New CI with https://github.com/ros2/ros2cli/pull/643:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14717)](http://ci.ros2.org/job/ci_linux/14717/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9469)](http://ci.ros2.org/job/ci_linux-aarch64/9469/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12384)](http://ci.ros2.org/job/ci_osx/12384/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14863)](http://ci.ros2.org/job/ci_windows/14863/)

audrow commented:
Thanks @rob-clarke for the PR! 🎉
